src/Functions/sleep.h

```diff
@@ -128,7 +134,19 @@ class FunctionSleep : public IFunction
                         "The maximum sleep time is {} microseconds. Requested: {} microseconds per block (of size {})",
                         max_microseconds, microseconds, size);
 
-                sleepForMicroseconds(microseconds);
+                while (microseconds)
+                {
+                    UInt64 sleep_ms = microseconds;
+                    if (query_status)
+                        sleep_ms = std::min(sleep_ms, /* 1 second */ static_cast<size_t>(1000000));
+
+                    sleepForMicroseconds(sleep_ms);
+                    microseconds -= sleep_ms;
+
+                    if (query_status && !query_status->checkTimeLimit())
+                        break;
+                }
+
                 ProfileEvents::increment(ProfileEvents::SleepFunctionCalls, count);
```

Comment:
The `ProfileEvents` events need to be adjusted now. `microseconds` has been changed in the loop and will always be zero, except with timeout+BREAK (which needs to be taken into account now too).